From 2f8dd45643ca9ea7124eec6e30d2e75118d30e36 Mon Sep 17 00:00:00 2001
From: Akira Ajisaka <aajisaka@apache.org>
Date: Thu, 23 Jun 2016 13:25:25 +0900
Subject: [PATCH 1852/2748] MAPREDUCE-6641. TestTaskAttempt fails in trunk.
 Contributed by Haibo Chen. (ported lambda to
 Supplier because CDH5 uses java7)

Change-Id: I58eca46044cd282b0fda448006eaac9c91647f58
---
 .../mapreduce/v2/app/job/impl/TestTaskAttempt.java |   16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/job/impl/TestTaskAttempt.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/job/impl/TestTaskAttempt.java
index 3ceb07a..8e1a95a 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/job/impl/TestTaskAttempt.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/job/impl/TestTaskAttempt.java
@@ -18,6 +18,7 @@
 
 package org.apache.hadoop.mapreduce.v2.app.job.impl;
 
+import static org.apache.hadoop.test.GenericTestUtils.waitFor;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
@@ -27,6 +28,7 @@
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
+import com.google.common.base.Supplier;
 import java.io.IOException;
 import java.net.InetSocketAddress;
 import java.util.HashMap;
@@ -390,7 +392,7 @@ private void testMRAppHistory(MRApp app) throws Exception {
   }
 
   private void testTaskAttemptAssignedKilledHistory
-      (FailingAttemptsDuringAssignedMRApp app) throws Exception {
+      (final FailingAttemptsDuringAssignedMRApp app) throws Exception {
     Configuration conf = new Configuration();
     Job job = app.submit(conf);
     app.waitForState(job, JobState.RUNNING);
@@ -400,8 +402,16 @@ private void testMRAppHistory(MRApp app) throws Exception {
     Map<TaskAttemptId, TaskAttempt> attempts = task.getAttempts();
     TaskAttempt attempt = attempts.values().iterator().next();
     app.waitForState(attempt, TaskAttemptState.KILLED);
-    Assert.assertTrue("No Ta Started JH Event", app.getTaStartJHEvent());
-    Assert.assertTrue("No Ta Killed JH Event", app.getTaKilledJHEvent());
+    waitFor(new Supplier<Boolean>() {
+      @Override public Boolean get() {
+        return app.getTaStartJHEvent();
+      }
+    }, 100, 800);
+    waitFor(new Supplier<Boolean>() {
+      @Override public Boolean get() {
+        return app.getTaKilledJHEvent();
+      }
+    }, 100, 800);
   }
 
   static class FailingAttemptsMRApp extends MRApp {
-- 
1.7.9.5

